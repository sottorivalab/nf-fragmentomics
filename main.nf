#!/usr/bin/env nextflow
/*
========================================================================================
FRAGMENTOMICS PIPELINE
========================================================================================
*/

nextflow.enable.dsl = 2

include { FRAGMENTOMICS } from './workflows/fragmentomics.nf'

def create_target_channel(LinkedHashMap row) {
    def meta = [
        name: row.name,
        source: row.source
    ]
    return [meta, file(row.bed)]
}

def create_sample_channel(LinkedHashMap row) {
    // create all at once
    def meta = [
        caseid: row.caseid,
        sampleid: row.sampleid,
        timepoint: row.timepoint
    ]

    // bam, bai and bw can be null
    return [
        meta, 
        row.bam ? file(row.bam) : null, 
        row.bai ? file(row.bai) : null, 
        row.bw  ? file(row.bw) : null
    ]
}

// MAIN WORKFLOW
workflow {
    
    main:
    // Init param files
    genome_2bit = params.genome_2bit ? Channel.fromPath(params.genome_2bit) : Channel.empty()
    blacklist_bed = params.blacklist_bed ? Channel.fromPath(params.blacklist_bed) : Channel.empty()

    // samples channel
    sample_ch = Channel.fromPath(params.input)
        .splitCsv(header:true, sep:',')
        .map{ create_sample_channel(it) }

    // targets channel
    target_ch = Channel.fromPath(params.targets)
        .splitCsv(header: true, sep:',')
        .map{ create_target_channel(it) }
        
    // filter targets for lines if not stubrun
    if (workflow.stubRun == false) {
        target_ch = target_ch
            .filter{ it ->
                it[1].readLines().size() > 1
            }
    }

    FRAGMENTOMICS(
        sample_ch,
        target_ch,
        genome_2bit,
        blacklist_bed
    )    

    // collect versions in a single file simple mode
    FRAGMENTOMICS.out.versions
        .unique()
        .map { version_file ->
            version_file.text
        }
        .unique()
        .view()
}
