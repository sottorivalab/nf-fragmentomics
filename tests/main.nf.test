
nextflow_function {

    name "Test Functions"
    script "main.nf"



    test("Test Function create_target_channel") {

        function "create_target_channel"

        when {
            function {
                """
                input[0] = ["name":"elk4vc", "source":"test", "bed":"$baseDir/tests/input/stub/elk4.bed"]
                """
            }
        }

        then {
            assert function.success
            assert snapshot(function.result).match()
        }

    }



    test("Test Function create_sample_channel") {

        function "create_sample_channel"

        when {
            function {
                """
                input[0] = ["caseid":"case1", "sampleid":"case1_T1", "timepoint":"T1", "bw":"$baseDir/tests/input/samplesheet_generator/bw/case1_T1.bw"]
                """
            }
        }

        then {
            assert function.success
            assert snapshot(function.result).match()
        }

    }



}