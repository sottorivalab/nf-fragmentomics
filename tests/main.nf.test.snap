{
    "Test Function create_target_channel": {
        "content": [
            [
                "test",
                "elk4.bed:md5,d41d8cd98f00b204e9800998ecf8427e"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-02-18T12:24:27.612498"
    },
    "Test Function create_sample_channel": {
        "content": [
            [
                {
                    "caseid": "case1",
                    "sampleid": "case1_T1",
                    "timepoint": "T1"
                },
                null,
                null,
                "case1_T1.bw:md5,d41d8cd98f00b204e9800998ecf8427e"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.3"
        },
        "timestamp": "2025-01-22T14:40:20.880655"
    }
}